libparsec/crates/client/src/client/recovery_device.rs

```diff
@@ -0,0 +1,373 @@
+            CertifPollServerError::Offline => ExportRecoveryDeviceError::Offline,
+            CertifPollServerError::InvalidCertificate(err) => {
+                ExportRecoveryDeviceError::InvalidCertificate(err)
+            }
+            CertifPollServerError::Internal(err) => err
+                .context("Cannot poll server for new certificates")
+                .into(),
+        })?;
+    Ok((passphrase, data))
+}
+
+pub async fn create_device_from_recovery(
+    cmds: Arc<AuthenticatedCmds>,
+    recovery_device: &LocalDevice,
+    device_label: &DeviceLabel,
+    save_strategy: DeviceSaveStrategy,
+    config_dir: PathBuf,
+) -> Result<AvailableDevice, ImportRecoveryDeviceError> {
+    let new_device =
+        LocalDevice::from_existing_device_for_user(recovery_device, device_label.clone());
+    // save recovery device
+    // TODO: save after upload
+    let access = {
+        let key_file = get_default_key_file(&config_dir, &new_device.device_id);
+        save_strategy.into_access(key_file)
+    };
+    let saved_device = save_device(&config_dir, &access, &new_device).await?;
+    let outcome = register_new_device(cmds, new_device, recovery_device.clone().into()).await?;
+
+    match outcome {
+        CertificateBasedActionOutcome::LocalIdempotent => return Ok(saved_device), // not sure how it could happen though
+        CertificateBasedActionOutcome::Uploaded {
+            certificate_timestamp,
+        }
+        | CertificateBasedActionOutcome::RemoteIdempotent {
+            certificate_timestamp,
+        } => PerTopicLastTimestamps::new_for_common_and_shamir(
+            certificate_timestamp,
+            certificate_timestamp,
+        ),
+    };
+
+    Ok(saved_device)
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ExportRecoveryDeviceError {
+    #[error("Component has stopped")]
+    Stopped,
+    #[error("Cannot reach the server")]
+    Offline,
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+    #[error("User {0} is revoked")]
+    UserRevoked(UserID),
+
+    #[error("Our clock ({client_timestamp}) and the server's one ({server_timestamp}) are too far apart")]
+    TimestampOutOfBallpark {
+        server_timestamp: DateTime,
+        client_timestamp: DateTime,
+        ballpark_client_early_offset: f64,
+        ballpark_client_late_offset: f64,
+    },
+    #[error(transparent)]
+    InvalidCertificate(#[from] Box<InvalidCertificateError>),
+    #[error(transparent)]
+    DataError(#[from] libparsec_types::DataError),
+    #[error(transparent)]
+    GetCertificateError(#[from] GetCertificateError),
+    #[error(transparent)]
+    CertifDeviceError(#[from] CertifRegisterNewDeviceError),
+    #[error(transparent)]
+    PlatformExportRecoveryDeviceError(#[from] PlatformExportRecoveryDeviceError),
+    #[error(transparent)]
+    ConnectionError(#[from] ConnectionError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ImportRecoveryDeviceError {
```

Comment:
Call it `ClientImportRecoveryDeviceError` since it will be used when calling `Client::import_recovery_device`